tools/themis/src/rules.rs

```diff
@@ -25,99 +23,83 @@ pub fn has_publish_spec(workspace: &Workspace) -> anyhow::Result<()> {
-        });
-    }
-
-    Ok(())
-}
-
-/// Ensure all crates are versioned to v0.0.0
-pub fn is_unversioned(workspace: &Workspace) -> anyhow::Result<()> {
-    let outliers = workspace
-        .members
-        .iter()
-        .filter(|pkg| {
-            !pkg.parsed.metadata["workspaces"]["independent"].as_bool().unwrap_or(false)
-                && pkg.parsed.version != semver::Version::new(0, 0, 0)
-        })
-        .map(|pkg| Outlier {
-            path: pkg.parsed.manifest_path.clone(),
-            found: Some(pkg.parsed.version.to_string()),
-        })
-        .collect::<Vec<_>>();
-
-    if !outliers.is_empty() {
-        bail!(ComplianceError {
-            msg: "These packages shouldn't be versioned".to_string(),
-            expected: Some(Expected { value: "0.0.0".to_string(), reason: None }),
-            outliers,
-        });
-    }
-
-    Ok(())
-}
-
-/// Ensures all crates have a rust-version spec less than
-/// or equal to the version defined in rust-toolchain.toml
-pub fn has_debuggable_rust_version(workspace: &Workspace) -> anyhow::Result<()> {
     let rust_toolchain =
         utils::parse_toml::<toml::Value>(&workspace.root.join("rust-toolchain.toml"))
             .context("Failed to read rust-toolchain file")?;
     let rust_toolchain = rust_toolchain["toolchain"]["channel"].as_str().unwrap().to_owned();
-
     let rust_toolchain = match semver::Version::parse(&rust_toolchain) {
         Ok(rust_toolchain) => rust_toolchain,
         Err(err) => {
-            utils::warn!(
+            bail!(
                 "semver: unable to parse rustup channel from {}: {}",
                 style::highlight("rust-toolchain.toml"),
                 err
             );
-
-            return Ok(());
         }
     };
 
-    let mut outliers = vec![];
+    let mut no_rust_version = vec![];
+    let mut wrong_rust_version = vec![];
+
     for pkg in &workspace.members {
         match &pkg.parsed.rust_version {
             Some(rust_version) => {
                 if !rust_version.matches(&rust_toolchain) {
-                    outliers.push(Outlier {
+                    wrong_rust_version.push(Outlier {
                         path: pkg.parsed.manifest_path.clone(),
                         found: Some(format!("{}", rust_version)),
                     });
                 }
             }
-            None => {
-                // we can skip, since we have has_rust_version check
-                continue;
-            }
+            None => no_rust_version
+                .push(Outlier { path: pkg.parsed.manifest_path.clone(), found: None }),
         }
     }
 
-    if !outliers.is_empty() {
+    if !no_rust_version.is_empty() {
+        bail!(ComplianceError {
+            msg: "These packages should specify a Minimum Supported Rust Version (MSRV)"
+                .to_string(),
+            expected: None,
+            outliers: no_rust_version,
+        });
+    }
+    if !wrong_rust_version.is_empty() {
```

Comment:
Let's remove `wrong_rustc_version` case altogether. I believe this is dead code: if it is not-empty, it means that we have some package whose rust-version is larger than that in rust-toolchain. But this will just fail compilation anyway: 

```
λ cargo check
error: package `near-o11y v0.0.0 (/home/matklad/p/nearcore/core/o11y)` cannot be built because it requires rustc 1.92.0 or newer, while the currently active rustc version is 1.64.0
```

Reply:
Agree, I have reworked the PR.